#11252 - Add checks as a pre-commit hook

jmarrec opened the pull request:
Pull request overview
---------------------

- Fixes #11252 

### Description of the purpose of this PR

* Adding a .pre-commit-config.yaml to register some common hooks (black, isort, trailing whitespace, etc) and our own scripts

Refactoring **all** of the checks we have in scripts/dev to generally do a few things:


* Dry-out code by reusing helpers
* Accept a positional `files` (nargs) argument so we can run checks ONLY on changed (staged) files instead of running on the full repo if it can be avoided (some stuff like check_for_enum_scope_usage has to run on all files)
     * If files isn't passed, it defaults to figuring out the list of files it needs to check, like before
* Parallelize the check accross a ProcessPoolExecutor
     * pre-commit will "chunk" the files as needed, but this is NOT what we want here because
          1. We still want to be able to run fast when we invoke the script directly (or via custom_check.sh as before)
          2. Because of GHA reporting: want one table per check, not several
* I have standardized the Error/Warning classes, in the helpers. And Added functionality to detect when we run on GHA and create actual annotations and a clear GITHUB_STEP_SUMMARY. You can see that in action on my fork at https://github.com/jmarrec/EnergyPlus/actions/runs/18731110985?pr=16#summary-53428133912

Currently I've made it so that autolint fixes are pushed when it runs on develop (note that most of our custom checks do not have a FIX, just a fail, so this mostly applies to the trailing whitespace, black, isort etc) 


On any modern machine, this is MUCH faster than the previous implementation. On my M1 mac, 10 cores, this takes 11s to do `pre-commit run --all-files`. And the pre-commit hook does more things than custom_check (black, isort, clang-format too, etc)

If I include "check_for_enum_scope_usage" which is not part of the regular pre-commit hook because it's still too slow,

`time pre-commit run --verbose --hook-stage manual --all-files`takes 23s.

In most cases you are changing a handful of files at once, so you should barely notice the pre-commit hook running as it'll take <2 s.


### Pull Request Author

<!-- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

<!-- This will not be exhaustively relevant to every PR. -->

 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
[chore] Run isort and black with 120 lines accross entire repo

```shell
isort --skip-glob='third_party/*' -l 120 .
black --exclude='third_party/*' -l 120 .
```

jmarrec committed:
Add a pyproject.toml to configure black & isort + a pre.commit config

jmarrec committed:
Adjust ignores

jmarrec committed:
[chore] pre-commit run --all-files trailing-whitespace

jmarrec committed:
transition_coil_heating was using python 2 print syntax and black choked on it

jmarrec committed:
Adjust config to exclude idf and co from end-of-file-fixer

jmarrec committed:
[chore] pre-commit run --verbose --all-files end-of-file-fixer

jmarrec committed:
Configure black and isort in pyproject.toml to also ignore bin/ and third_party/

jmarrec committed:
Getting ready to do some large refactor

jmarrec committed:
Make a hook and use parallelizationf or check_constexpr, runs in 1s versus 10s on all files

jmarrec committed:
Do validate idd units

jmarrec committed:
Update base_hook

jmarrec committed:
Wrong format for precommit

jmarrec committed:
Use official pre-commit hook fix-byte-order: TODO: probably delete check_for_bom_in_idfs

Some of these MD files have a ton of BOM marker

design/FY2019/NFP-Connect Split DX to VRF Refrigerant Loop.md has 18 (eighteen!) of them

jmarrec committed:
Do verify-idfs-in-cmake

jmarrec committed:
Do check non UTF8

jmarrec committed:
do check-stray-fields-in-idd

jmarrec committed:
Do find_byref_bool_override.py

jmarrec committed:
Do license files

jmarrec committed:
Delete verify_file_encodings (check-non-utf8). Modernize check_for_bom though I might delete it

jmarrec committed:
Do check for tabs

jmarrec committed:
Parallelize check_for_enum_scope_usage: 1min20 before, 14.5s now (16 threads here)

jmarrec committed:
Disable check-for-enum-scope-usage in precommit: takes too long

jmarrec committed:
Botched the license-check extensions (forgot .hh)

jmarrec committed:
Do find-included-cc-files

jmarrec committed:
Do analyze state

jmarrec committed:
Do check-for-c-style-comments

jmarrec committed:
Do check-for-malformed-enums and check-for-switch-case-parentheses

jmarrec committed:
Finish with check_format_strings

jmarrec committed:
Run mypy and fix all issues. Most of them are just static typing, but a few would be runtime errors

jmarrec committed:
Test it up live and install tabulate for the GITHUB_STEP_SUMMARY

jmarrec committed:
Tweak

jmarrec committed:
if always

jmarrec committed:
Style errors in step summary with icons + downgrade unmatched parenthesis to Info so it's not displayed

The offending files have constexpr string_view with numbered lists like "1) blabla"

jmarrec committed:
Alright, go with pre-commit now

jmarrec committed:
Add clang format hook (pulls binaries from pypi) and register check-for-enum-scope-usage as a 'manual' stages

jmarrec committed:
Run also check_for_enum_scope_usage via pre-commit

jmarrec committed:
Commit the pre-commit changes on push

jmarrec committed:
Remove build on push for my branch + reorder stages and tweak

jmarrec committed:
pre-commit run --all-files after rebase

jmarrec committed:
Add documentation on the pre-commit

jmarrec commented:
I know this looks scary to review due to the number of changes here.

I have a branch where I have reduced this into 3 commits:
* One: make the tooling changes
* Two & three: apply all checks

https://github.com/NREL/EnergyPlus/compare/11252_precommit-squashed?expand=1

If you prefer I can replace the PR with that one.


Anyways, the actual code changes are in this commit: https://github.com/NREL/EnergyPlus/commit/c50b0f451191ff1ac4b1cea5215a637240da0f69

mitchute committed:
Merge branch 'develop' into 11252_precommit

mitchute committed:
fix botched merge

mitchute committed:
more merge issues

mitchute committed:
run pre-commit

mitchute commented:
Pretty cool. Pretty, pretty cool.

This is going to be a great addition to our workflows, though I suspect we'll all need some education on how to work with it. We'll hold this until after the release in an case.

I installed the dependencies manually like you show in the [GHA workflow](https://github.com/NREL/EnergyPlus/pull/11286/files#diff-02696ee1a7985368fc452a3041c62445e176a5c7ccca396e37da88a39b99780cR22), and that worked, but installing via the `pyproject.toml` file didn't seem to work. Maybe I did that part wrong. 

```sh
(venv) mmitchel-39518s:EnergyPlus mmitchel$ pip install -e .[dev]
Obtaining file:///Users/mmitchel/Projects/EnergyPlus
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
WARNING: energyplus 25.2.0.1 does not provide the extra 'dev'
Building wheels for collected packages: energyplus
  Building editable for energyplus (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building editable for energyplus (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [90 lines of output]
      /private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/dist.py:289: UserWarning: Unknown distribution option: 'build_base'
        warnings.warn(msg)
      running editable_wheel
      creating /private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info
      writing /private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info/PKG-INFO
      writing dependency_links to /private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info/dependency_links.txt
      writing top-level names to /private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info/top_level.txt
      writing manifest file '/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info/SOURCES.txt'
      reading manifest file '/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info/SOURCES.txt'
      adding license file 'LICENSE.txt'
      writing manifest file '/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus.egg-info/SOURCES.txt'
      creating '/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus-25.2.0.1.dist-info'
      creating /private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-ephem-wheel-cache-g8krv1f4/wheels/91/85/1a/d36ebfd5cc98b9fa4d6baf0e9455cb2e88fe5d7a6db901805f/tmpgbll_m8y/.tmp-eei_me2k/energyplus-25.2.0.1.dist-info/WHEEL
      running build_ext
      -- Using CCACHE
      -- /opt/homebrew/bin/ccache found and enabled
      CMake Error at CMakeLists.txt:19 (message):
        CMAKE_OSX_DEPLOYMENT_TARGET not set.  Please set
        CMAKE_OSX_DEPLOYMENT_TARGET to 10.15 or greater and try again.
      
      
      -- Configuring incomplete, errors occurred!
      Traceback (most recent call last):
        File "/Users/mmitchel/Projects/EnergyPlus/venv/lib/python3.14/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
          main()
          ~~~~^^
        File "/Users/mmitchel/Projects/EnergyPlus/venv/lib/python3.14/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main
          json_out["return_val"] = hook(**hook_input["kwargs"])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/mmitchel/Projects/EnergyPlus/venv/lib/python3.14/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 303, in build_editable
          return hook(wheel_directory, config_settings, metadata_directory)
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/build_meta.py", line 468, in build_editable
          return self._build_with_temp_dir(
                 ~~~~~~~~~~~~~~~~~~~~~~~~~^
              cmd, ".whl", wheel_directory, config_settings
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          )
          ^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/build_meta.py", line 404, in _build_with_temp_dir
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/build_meta.py", line 512, in run_setup
          super().run_setup(setup_script=setup_script)
          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/build_meta.py", line 317, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 214, in <module>
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/__init__.py", line 115, in setup
          return distutils.core.setup(**attrs)
                 ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/core.py", line 186, in setup
          return run_commands(dist)
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/core.py", line 202, in run_commands
          dist.run_commands()
          ~~~~~~~~~~~~~~~~~^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/dist.py", line 1002, in run_commands
          self.run_command(cmd)
          ~~~~~~~~~~~~~~~~^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/dist.py", line 1102, in run_command
          super().run_command(command)
          ~~~~~~~~~~~~~~~~~~~^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/dist.py", line 1021, in run_command
          cmd_obj.run()
          ~~~~~~~~~~~^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/command/editable_wheel.py", line 139, in run
          self._create_wheel_file(bdist_wheel)
          ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/command/editable_wheel.py", line 349, in _create_wheel_file
          files, mapping = self._run_build_commands(dist_name, unpacked, lib, tmp)
                           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/command/editable_wheel.py", line 272, in _run_build_commands
          self._run_build_subcommands()
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/command/editable_wheel.py", line 299, in _run_build_subcommands
          self.run_command(name)
          ~~~~~~~~~~~~~~~~^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/cmd.py", line 357, in run_command
          self.distribution.run_command(command)
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/dist.py", line 1102, in run_command
          super().run_command(command)
          ~~~~~~~~~~~~~~~~~~~^^^^^^^^^
        File "/private/var/folders/lg/t8p8n3551qdd02nqp0qbc2_cdz92s2/T/pip-build-env-kaleh33a/overlay/lib/python3.14/site-packages/setuptools/_distutils/dist.py", line 1021, in run_command
          cmd_obj.run()
          ~~~~~~~~~~~^^
        File "<string>", line 171, in run
      Exception: CMake failed to configure EnergyPlus, check error logs, raw error message: Command '['cmake', '-G', 'Unix Makefiles', '-DBUILD_FORTRAN=OFF', '-DCMAKE_BUILD_TYPE=Release', '/Users/mmitchel/Projects/EnergyPlus']' returned non-zero exit status 1.
      An error occurred when building editable wheel for energyplus.
      See debugging tips in: https://setuptools.pypa.io/en/latest/userguide/development_mode.html#debugging-tips
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building editable for energyplus
Failed to build energyplus
error: failed-wheel-build-for-install

× Failed to build installable wheels for some pyproject.toml based projects
╰─> energyplus
```

mitchute commented:
In case people want to see what this looks like.

```sh
(venv) mmitchel-39518s:EnergyPlus mmitchel$ pre-commit run --all-files
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/pycqa/isort.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-clang-format.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pycqa/isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-clang-format.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing testfiles/SolarHeatingPythonPlugin.py

fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing testfiles/SolarHeatingPythonPlugin.py

check for merge conflicts................................................Passed
check yaml...............................................................Passed
fix utf-8 byte order marker..............................................Passed
black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted cmake/PythonFixUpTclTk.py
reformatted idd/schema/modify_schema.py
reformatted testfiles/PythonPlugin_SingleFamilyHouse_TwoSpeed_MultiStageElectricSuppCoil.py
reformatted testfiles/SolarHeatingPythonPlugin.py

All done! ✨ 🍰 ✨
4 files reformatted, 123 files left unchanged.

isort....................................................................Failed
- hook id: isort
- files were modified by this hook

Fixing /Users/mmitchel/Projects/EnergyPlus/testfiles/SolarHeatingPythonPlugin.py

clang-format.............................................................Passed
check constexpr usage....................................................Passed
license check............................................................Passed
validate IDD units.......................................................Passed
check stray fields in IDD................................................Passed
verify IDFs in testfiles CMake...........................................Passed
check files are UTF-8....................................................Passed
verify cmake dirs........................................................Passed
find byref bool override.................................................Passed
check for tabs in IDFs...................................................Passed
find included cc files...................................................Passed
analyze state............................................................Passed
check for C style comments...............................................Passed
check for switch case parentheses........................................Passed
check for malformed enums................................................Passed
check format strings.....................................................Passed
```
```sh
(venv) mmitchel-39518s:EnergyPlus mmitchel$ git status
On branch 11252_precommit
Your branch is up to date with 'origin/11252_precommit'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   cmake/PythonFixUpTclTk.py
        modified:   idd/schema/modify_schema.py
        modified:   testfiles/PythonPlugin_SingleFamilyHouse_TwoSpeed_MultiStageElectricSuppCoil.py
        modified:   testfiles/SolarHeatingPythonPlugin.py

no changes added to commit (use "git add" and/or "git commit -a")
```
```sh
(venv) mmitchel-39518s:EnergyPlus mmitchel$ pre-commit run --all-files
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
check for merge conflicts................................................Passed
check yaml...............................................................Passed
fix utf-8 byte order marker..............................................Passed
black....................................................................Passed
isort....................................................................Passed
clang-format.............................................................Passed
check constexpr usage....................................................Passed
license check............................................................Passed
validate IDD units.......................................................Passed
check stray fields in IDD................................................Passed
verify IDFs in testfiles CMake...........................................Passed
check files are UTF-8....................................................Passed
verify cmake dirs........................................................Passed
find byref bool override.................................................Passed
check for tabs in IDFs...................................................Passed
find included cc files...................................................Passed
analyze state............................................................Passed
check for C style comments...............................................Passed
check for switch case parentheses........................................Passed
check for malformed enums................................................Passed
check format strings.....................................................Passed
```

mitchute committed:
Merge branch 'develop' into 11252_precommit

mitchute committed:
run pre-commit

jmarrec committed:
Merge branch 'develop' into 11252_precommit

jmarrec commented on testfiles/SteamSystemAutoSize_DistrictHeatingSteam.idf:
```diff
@@ -3509,6 +3509,29 @@
+
+  Output:Variable,SPACE5-1 Reheat Coil Steam Outlet, System Node Temperature,hourly;
+
+
```
I was puzzled for a while as to why two integration tests had audit diffs with NumConsideredVariables.

I think it's a botched merged commit in b99c3bfa487cb682be8ff258db4558eca15a320c

jmarrec replied:
```diff
- NumConsideredOutputVariables=          38
+ NumConsideredOutputVariables=          49
```

jmarrec replied:
Hopefully 37c22e8f824b8e9ef394c8ce6318c05b79e78169 is enough to clean no diffs anymore!

jmarrec committed:
Undo wrong merge changes duplicating OutputVariables from b99c3bfa487cb682be8ff258db4558eca15a320c

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 11252_precommit

mitchute committed:
changes needed for running on windows

jmarrec committed:
Another encoding=utf-8 needed for windows

jmarrec committed:
Use language: python so that it uses the venv python

Otherwise with @mitchute's changes it not longer works from my git client (GitKraken)

The issue is that for entry:
* If we use "python3", it doesn't work on windows (C:/Python312/python.exe) exists only
* If we use "python", it works on Ubuntu when I commit from the terminal because I have my virtualenv loaded and that has a shim, but it doesn't work when I commit from GitKraken, because I don't have the virtualenv loaded

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 11252_precommit

jmarrec commented:
> Gah, didn't mean to push 94dae5fddee0fe531acefa61bfd61dcbd834987c 😒 

Reset locally before and force push? I've got no issue with that (I personally strongly encourage keeping history clean)

mitchute commented:
OK, thanks @jmarrec! This will be a nice addition to our development checks. We're ready to go here. 

For everyone else, the process for getting this set up and running this locally is outlined under [scripts/dev/README.md](https://github.com/NREL/EnergyPlus/pull/11286/files#diff-047e1211d0f666d58aab574207023ab357c53dc5cf8080b10b5368986dbb24ac), but I can help people get set up, as needed, if you have issues.

mitchute reviewed (approved):
Approving here. I've done lots of testing locally on Windows, Linux, and Mac. This is working via the command line and via git clients like GitHub's desktop app and TortoiseGit.
